File: 0875-koko-eating-bananas/0875-koko-eating-bananas.cpp
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        
        int lo = 1;
        int hi = *max_element(piles.begin(), piles.end());
        
        int ans = hi;
        
        while (lo <= hi) {
            int mid = lo + (hi - lo) / 2;
            
            long long totalHour = 0;
            
            // calculate total hours for this speed
            for (int i = 0; i < piles.size(); i++) {
                totalHour +=ceil((double)piles[i] / mid);
  
            }
            
            if (totalHour <= h) {
                ans = mid; 
                hi = mid - 1;   
            }
            else {
                lo = mid + 1;  
            }
        }
        
        return ans;
    }
};

File: 0875-koko-eating-bananas/0875-koko-eating-bananas.py
import math
from typing import List

class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        l = 1
        r = max(piles)
        ans = r
        
        while l <= r:
            mid = (l + r) // 2
            hours = 0
            for p in piles:
                hours += math.ceil(p / mid)
            if hours <= h: 
                ans = mid       
                r = mid - 1     
            else:
                l = mid + 1   
                
        return ans

